Pattern: InnerSource License (#147)
gruetter/innerSourceTest: innersource-license.md

## Title

InnerSource License

## Patlet

> Two companies that belong to the same group want to share software source code with each other but they are concerned about the implications in terms of legal liabilities or cross-company accounting. An **InnerSource License** provides a reusable legal framework for the inter-company agreement within the group, which opens up new collaboration options.

Reviewer: ```suggestion
An organization faces legal uncertainty about InnerSource collaboration among the multiple legal entities its composed of. An **InnerSource License** makes explicit the rights and obligations of those a legal entities providing and using the component.```

Reviewer: This is technically two proposed changes in one - both open for discussion obviously:

**First)** I'd propose to use `organization` composed of multiple `legal entities`:

- *Organization* - an organization in a broad and less narrowly defined sense (e.g. Lufthansa)
- *Legal entity* - an entity that has own legal rights and obligations (e.g. Lufthansa Systems GmbH, Lufthansa Industry Solutions TS GmbH, ...)

My worry is that both `company` and `group` are highly ambiguous terms. Readers might have different ideas of what a group or company is. 

Another I had in mind was to borrow terminology from tax accounting and talk about `associated enterprises` and plainly other, non-associated enterprises. However, this doesn't help us in distinguishing the fine-grained parts (legal entities) from the whole (organization).

**Second)** I adapted (shortened) the patlet to be more focused on what - I think - are the key insights of the pattern.

Reviewer: (I'll not point out the first suggestion again to not clutter the PR ... I am happy to provide commit suggestions once we read a consensus though)

Author: Cornelius and I also went back and forth between calling it *enterprise* or *group*. And we landed at *group* eventually.

For the time being I suggest to keep it as *group* so that we don't have to reword everything. Also I don't know which terms are easier to understand, or which ones are more correct in legal speak.

Maybe when other companies that have similar challenges add their experience to this pattern, then we can test which term they prefer?

Author: I really like the edits you did to the Patlet!
I will take parts of that but keep the group/company terms as they are for the time being.

Author: I was only able to work in part of that, without changing the original meaning of the Patlet too much.

Will keep this conversation open also for reference for Cornelius.

Reviewer: > For the time being I suggest to keep it as group so that we don't have to reword everything.

I know writing is a lot of work and time is a scarce resource. Let's still try to find the best terminology possible. We / I can support with the rewording once we settled on terms.

> Maybe when other companies that have similar challenges add their experience to this pattern, then we can test which term they prefer?

In the organizations I worked with, we ended up using exclusively `legal entity` to talk about the - well - legal entities.

Reviewer: Summary of arguments, both gathered from this discussion and from my head:

`group, company`
- [PLUS] Fits well with the proposed pattern (because it's the terms the author's had in mind initially)
- [PLUS] Less re-writing burden
- [MINUS] Excludes specific organizations that are not groups/companies but still consist of multiple legal entities and want to do InnerSource (private / publich research institutions, government agencies, ...)
- [MINUS] Ambiguity of both terms
- [MINUS] Less clear on describing the actual type of building block of the organization that is relevant for this pattern (which, as I read it, applies primarily to multiple legal entities)

So, my suggestion would still be, to use different terms, probably `organization` and `legal entity`

Reviewer 2: > So, my suggestion would still be, to use different terms, probably organization and legal entity

I like the term `legal entity` because that's closely tied to the need of having legal agreements in place. `organization` as a general term would also work for me. No matter what terms we use, we should make sure that we have a definition somewhere which explains them in the context of this pattern.

Reviewer: > No matter what terms we use, we should make sure that we have a definition somewhere which explains them in the context of this pattern.

+1000! Let's add those definitions to the context section

Reviewer 3: I'll go with @cornelius point of "what ever we use we should make sure that we have a definition somewhere". Let's use a glossary for that one. We argued that this would be useful across patterns in any case. Starting it can be as simple as creating a simple text file with "term: explanation" pairs in it and link to that, it's hypertext after all. 
We can always fancy that up later.

Other than that: If we talk about legal agreements between legal entities it might not harm calling them that. We could give an example "A legal entity in this context can be x,y,z". But effectively the point is definition of terms before usage and usage according to definition.

Author: Also see inline comment from @lenucksi:
https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/147/files#r417457368

Author: Based on all the feedback I will rewrite the pattern to use **organization** (for group/enterprise) and **legal entity** (for company/subsidiary). I will put those as definitions to the Context section.

Will resolve this thread here once done.

Author: I have rewritten the pattern accordingly.
 
I have placed the glossary at the end of this file, as I cannot claim that this is how these terms are used in other patterns. If we really want to produce a global glossary, I suggest that we do that as a separate activity outside of this PR>

@maxcapraro would you mind reviewing this, as you brought this topic up?

My observations after doing the rewrite:

- the pattern is certainly more consistent now 👍 
- it sounds much more like a legal document, and less like it is written for humans (which I personally don't like) 👎 
- there are a couple of places where replacing the word "company" without "legal entity" would make the sentences longer,  and also a bit awkward. I have left those as they were. Examples:
  - cross-company accounting
  - Company culture
  - inter-company relationships

Still I think that this was a good iteration and if this gets sign-off, I would gladly get this merged to master as it already provides value to the InnerSource community as is.

> ## Problem

Reviewer: The section describes the problem of legal uncertainty really well. I feel there is a second (related) problem that the license solves: 

In addition to the legal uncertainty, there is the a psychological uncertainty: "There is this piece of code. Can I really just use this for my work?" The license resolves this as well. This is similar to what licenses in open source (and particularly in the context free software) achieved: They spread the gospel of collaboration.

Reviewer: (On the other side: This could distract readers from the key problem at hand - the legal requirements and the barriers they can introduce)

Reviewer: Yippie! That goes into the direction of what you observed at DB (further down): 

> One positive effect that is already showing is that it simplifies the conversation, especially if some of the involved parties don’t know the InnerSource concept that well yet. Licenses are a well-known concept, therefore having an InnerSource License is a great discussion starter.

Author: I am unclear if you are suggesting any change here?

Reviewer: My suggestion is for you, as authors, to consider whether the psychological uncertainty is something that you would want to add.

Reviewer 3: Judging from my previous work and discussions with engineers and other non-legal professionals I can very much confirm that "psychological unsafety" is a thing. 
However, I can only join @spier's question here:  "Are you suggesting a change"?
If so: I'd be happy about (a) commit suggestion(s) at the right place(s).

Author: For now, let's not add anything here.

Once other people/companies that are following this pattern start contributing to this, they can add **psychological uncertainty** here if that is something that they have seen in the context of this pattern.

> When two companies within a group want to share code with each other, they need an agreement about the terms and often a legal contract. Creating such agreements on a per project basis takes effort and creates a barrier for sharing. i.e. a team/company might decide not to share their source code with another company in the group because it seems complicated.

Reviewer: I'd like to dig a little deeper on the root causes for the `... they need ...` part and make them more explicit. What are the reasons?

- Is it a fear of illegal profit shifting between the legal entities?
- Is it a general accounting issue / obligation? (i.e. you need to make explicit rights and obligations to properly record liabilities and new assets that InnerSource-style sharing brings to the different legal entities)
- Is it something else?

Reviewer: (Digging could happen as part of the forces, too)

Author: The Patlet section contains

> they are concerned about the implications in terms of legal liabilities or cross-company accounting

@maxcapraro are you suggesting that we elaborate on that part in the Problem section more?

Reviewer 2: Regarding the motivation I think there are two high-level root causes:
* There are legal contracts in place which cover sharing source code but they are not standardized (so they create additional effort in negotiating and understanding for every project) or they don't allow sharing source code in an open enough sense as an Inner Source approach would suggest.
* There are no contracts in place but source code is shared informally. That might create uncertainty in cases where clarity about ownership and rights and obligations is needed.

Accounting, profit sharing, taxes are all things which could be covered in an Inner Source license, but they also could be left to other legal constructions.

Ownership of code, relicensing, liability, warranty are topics you probably would typically cover in a license.

In summary I would say the goal is to create clarity about the terms under which source code is shared.

Reviewer: > Accounting, profit sharing, taxes are all things which could be covered in an Inner Source license, but they also could be left to other legal constructions.
>
> Ownership of code, relicensing, liability, warranty are topics you probably would typically cover in a license.

I agree fully. This is how I have seen InnerSource licenses been used in organizations. For some organizations the tax / transfer pricing discussion is the first that makes explicit the need for a clear legal framework. Hence my question :) 

> - There are legal contracts in place which cover sharing source code but they are not standardized (so they create additional effort in negotiating and understanding for every project) or they don't allow sharing source code in an open enough sense as an Inner Source approach would suggest.
> 
> - There are no contracts in place but source code is shared informally. That might create uncertainty in cases where clarity about ownership and rights and obligations is needed.

I like this wording very much. I suggest to merge it into that section! :)

Reviewer 3: @cornelius or @maxcapraro Could you please turn said wording into a commit suggestion  or commit and have it merged? That way we have conserved it and ensure proper attribution.

Author: I have added the two bullets from @cornelius with slight modifications to the **Forces** section, as these are "constraints that can be changed at a cost".

Resolving this thread.

> 

Reviewer: Could be `When two or more ...` (Can't make commit suggestion because I do not want to mix the two threads)

Author: Great suggestion. Have added this in.

> Barriers for sharing can lead to silos and duplication of effort in rebuilding similar solutions in multiple parts of the group.

Reviewer: ```suggestion
This uncertainty creates an effective barrier to InnerSource collaboration among the different legal entities of the organization.
```

I feel the fact that the patterns in this repo are "InnerSource Patterns" establishes the context that we want to bring down silos, avoid duplication, forster reuse and collaboration. I'd suggest we not list them again to avoid duplication.

Author: Just like our other thread about the **InnerSource Context**, I would like to keep this in here to allow the readers to understand the pattern without having to read anything else.

@maxcapraro please close this conversation if you are fine with leaving this section as is.

> At the time of sharing the source code, it can not be reliably predicted what the value of sharing will be. Hence if the activity of sharing has any significant cost, the companies are unlikely to do it, as they are concerned about the return on investment.

Reviewer: This is an excellent description of why this problem is a problem 👍

Reviewer: Some sugestion for word smithing (feel free to commit, steal parts, or ignore)
```suggestion
At the time of sharing the source code, it can not be reliably predicted what the value of sharing will be. Hence if the activity of sharing creates significant effort (i.e. negotiating terms for the usage), the companies are less likely to do it: They cannot predict if the effort will pays off.
```

Author: Great feedback. I worked in something similar.

## Context

- A large group with many companies (subsidiaries) that want to share code. (when the group gets larger, the value of this pattern increases)

> - These companies are legally independent units.

Reviewer: ```suggestion
- These legal entities are entities with own legal rights, obligations
```

Author: I worked in something similar. Resolving.

- Multiple of these companies/subsidiaries are developing software, and are using services of the other companies. They have a motivation to contribute to each other’s source code.

> - Certain size of developers (crowdsourcing and ecosystem effects of InnerSource)

Reviewer: ```suggestion
```

Doesn't that apply to InnerSource in general? Is it a contextual factor only important for this pattern?

Reviewer: (Mental note to ourselves: Maybe we need to define a general context (some kind of an `AbstractSuperContext`) for all patterns)

Author: I agree that there are likely some general points that always apply when an InnerSource approach is considered by a company.

For the time being I would like to keep this in here though, as we don't have such a general **InnerSource Context** (`AbstractSuperContext`) description yet. My assumption is that the readers need to be able to "read the pattern on it's own". 

But I can see that if we ever want to turn all proven patterns into a book, there would be an intro section that describes the **InnerSource Context**.

Do you want to create an issue to describe that **InnerSource Context** (`AbstractSuperContext`)? @gruetter could be interested in that. At the summit the topic of "how large does a company have to be for InnerSource to make sense" came up in a breakout that we were in.

@maxcapraro please close this conversation if you are fine with leaving this section as is.

Reviewer 2: I think the specific condition for an Inner Source License is that the organization has a complexity so that there are legal contracts necessary within the organization. That probably corresponds with size and numbers of developers in most cases.

Reviewer: >  [...] AbstractSuperContext [...]

Agreed, @spier! That's a topic for elsewhere :)

> I think the specific condition for an Inner Source License is that the organization has a complexity so that there are legal contracts necessary within the organization. That probably corresponds with size and numbers of developers in most cases.

Cool. It doesn't seem to be merely the size then. A possible fix could mabye be something like that? :)

```suggestion
- A sufficient complexity of the organization and its organizational structure
```

## Forces

- **Level of effort** required to write formal agreements, especially if they need to take into account technical, legal, and business perspectives.

> - Enterprises have many **internal regulations,** any agreements have to comply with these regulations, e.g. security, privacy, procurement processes, etc. The volume of regulations can make it difficult to assess whether sharing software between two companies is compliant with these regulations, especially when there is no standard procedure.

Reviewer: `Internal regulation` are one thing. Another thing are plain legal requirements. Did you interviews yield any insights on this (potential) additional force?

Author: We did not go deeper into this. The **internal regulations** aspect seemed to be the more pressing one.

@cornelius what do you say, do **external regulations** also impact this pattern? Should we add a bullet about **external regulations** to the **Forces**?

Reviewer 2: I would say that external regulations do have an impact, but not a very specific one. Legal documents have to match plain legal requirements, but that's kind of a given.

Reviewer: > I would say that external regulations do have an impact, but not a very specific one. 

I assume you worked a lot with our legal team in designing your org's InnerSource license. Would you be open to asking them if they see specific external regulations that require that? (I understand if you don't want to: questions to the legal team can leverage unintended consequences).

If in your observation the internal regulations were the key force. Please feel free to resolve this comment @cornelius and @spier.

Author: Given that 10 days have passed, I will resolve this. Can be expanded on later by other companies/contributors.

- If any of the companies in the group has a **business model** that depends on proprietary code and accounting of licensing fees within the group

> - **Company culture** that isn’t used to sharing code

Reviewer: My first reaction was: "Well, that's a general InnerSource force, so maybe we should drop it." But the more I think about it, I think it was right to list it here: Missing routine with sharing / collaborating, might increase the need for a license.

Maybe we should add a sentence though, to explain how this force ties into the pattern?

Author: Do you mean the whole section you marked or only **Company culture**?

Reviewer: (Intended to only mark `Company culture`)

To clarify the suggestion: Make explicit how `company culture` plays into this pattern:
```suggestion
- **Company culture** that isn’t used to InnerSource collaboration and sharing code. This results in uncertainty about the rights and obligations when using shared code.
```

One could very well argue: "Whoops. Doesn't that drift too much into the problem section?" but I think it is important to have a tie in between the forces and other sections where possible.

Author: I will merge your suggestion in. This also points to the "psychological uncertainty" of the individual developer a little bit. So at least we have a note of it in here that we can elaborate on later if we want to.

- Freedom over using the software leads to competition, and spread of ownership

## Solutions

Creating an **InnerSource License** customized to the needs of the group in question (and their companies). This license needs to be generic enough to be applied to the most important inter-company relationships.

It is important to write the InnerSource License such that it truly allows for OpenSource-like collaborations across organizational boundaries. Therefore the 4 freedoms of free software should be integrated into the license.

The License is written as a formal legal document, and can be used as part of contracts between the subsidiaries to govern the code sharing agreements.

## Resulting Context

> With the InnerSource License, we have a tool to share code between companies within our group.

Reviewer: Can we beef that up somehow? 

- What are thing people should be aware of when applying this pattern?
- What were positive side effects observed?
- What were negative side effects observed? (I always feel that is the greatest part of the resulting context)

Author: I think what you are looking for is mostly described in **Known Instances**. There we tried to describe the observations that are specific to DB Systel. Also as they have barely rolled this InnerSource License concept out, the experience is not that hardened yet. 

Maybe we can review this **Resulting Context** again in a couple of months or together with other companies that have applied this pattern too?

I would not change this for now.

@maxcapraro please close this conversation if you are fine with leaving this section as is.

Reviewer: My guts feeling would be to still synthesize it and already add it to the resulting context. I'd like to hear @lenucksi's / other's opinion on that, though.

Reviewer 3: What @spier states here is actually what I would more or less add to the resulting context.
- We could very thinly paraphrase the Known Instances and point there for more
-  and also state that research is still out on specific,reliable resultung context results.

Author: I have beefed up the resulting context just a little bit, following @lenucksi suggestion. Resolving this thread.

> ## Known Instances

Reviewer: I konw of two other organizations using an MIT-based license. I'll contact them and ask if they would want to provide a paragraph about it here or at least allow us to write about it.

This shouldn't stop this PR from being merged though - we can always add it later.

Author: Input from other companies with similar models will certainly be interesting. Just as you said, let's move that out of the context of this PR though.

If you want, you can even file a new issue that says "Harden InnerSource License pattern with experience from x and y".

Will resolve this without action.

## Problem

When two or more companies within a group want to share code with each other, they need an agreement about the terms and often a legal contract. Creating such agreements on a per project basis takes effort and creates a barrier for sharing. i.e. a team/company might decide not to share their source code with another company in the group because it seems complicated.

Barriers for sharing can lead to silos and duplication of effort in rebuilding similar solutions in multiple parts of the group.

At the time of sharing the source code, it can not be reliably predicted what the value of sharing will be. If the activity of sharing requires significant effort (i.e. negotiating terms for the usage), the companies are less likely to do it, as they are concerned about the return on investment.

## Context

- A large group with many companies (subsidiaries) that want to share code. When the group gets larger, the value of this pattern increases.
- These companies are legal entities with their own legal rights and obligations.

> - Multiple of these companies/subsidiaries are developing software, and are using services of the other companies. They have a motivation to contribute to each other’s source code.

Reviewer 3: From reading this I feel that we might be able to streamline the terms here a bit. 
I read: 
- company == subsidiary (BSH, RBPT, ...)
- large group == the entire thing (Bosch) - this is what I would know as "Konzern" in German.

I think subsidiary is actually a good, pretty unambiguously word. I'd vote to use this instead of "company" after introducing the word subsidiary.

Author: I have added this feedback to the main thread about the terminology. 

Will resolve this individual comment now.

- A sufficient complexity of the organization and its organizational structure

## Forces

- **Level of effort** required to write formal agreements, especially if they need to take into account technical, legal, and business perspectives.

> - Enterprises have many **internal regulations,** any agreements have to comply with these regulations, e.g. security, privacy, procurement processes, etc. The volume of regulations can make it difficult to assess whether sharing software between two companies is compliant with these regulations, especially when there is no standard procedure.

Reviewer 3: Here, the term 'Enterprise' just emerged from the scrubs. 
Could we align it with the other terms the patterns has or explain the difference in semantics to the others?

Author: Have replaced it with "Large groups of companies (enterprises)" for now.

It might be that we revisit these terms completely but that is discussed at length in other comments already. Hence closing this.

- Multiple of these legal entities are developing software, and are using services of the other legal entities. They have a motivation to contribute to each other’s source code.
- A sufficient complexity of the organization and its organizational structure

## Forces

- **Level of effort** required to write formal agreements, especially if they need to take into account technical, legal, and business perspectives.
- A large organization (consisting of many legal entities) has many **internal regulations**. Any new agreements that are made have to comply with these regulations, e.g. security, privacy, procurement processes, etc. The volume of regulations can make it difficult to assess whether sharing software between two legal entities is compliant with these regulations, especially when there is no standard procedure.
- If any of the legal entities in the organization has a **business model** that depends on proprietary code and accounting of licensing fees within the organization
- **Company culture** that isn’t used to InnerSource collaboration and sharing code. This results in uncertainty about the rights and obligations when using shared code.

> - Freedom over using the software leads to competition, and spread of ownership

Reviewer 4: This one' a bit fuzzy to me. 

```suggestion
- Unregulated use of the software can lead to unclear ownership structures, duplication of improvement effort or unwanted competition
```

- There are legal contracts in place which cover the sharing of source code. These contracts are not standardized, so they create additional effort in negotiating and understanding for every project. The existing contracts may also not allow sharing source code in an open enough sense to support a true InnerSource approach.

> - Alternatively, there are no legal contracts in place but source code is shared informally. That might create uncertainty in cases where clarity about ownership and rights and obligations is needed.

Reviewer 4: This one is at least partially a duplication of the force above, regarding the consequences of _Freedom over using the software"_

## Solutions

Creating an **InnerSource License** customized to the needs of the organization in question (and their legal entities). This license needs to be generic enough to be applied to the most important inter-company relationships.

It is important to write the InnerSource License such that it truly allows for OpenSource-like collaborations across the boundaries of the involved legal entities. Therefore the 4 freedoms of free software should be integrated into the license.

The License is written as a formal legal document, and can be used as part of contracts between the legal entities to govern the code sharing agreements.

> 

Reviewer 4: We should mention that an InnerSource license needs to be ratified by the responsible bodies in the respective organisation. If there is no such ratification, the license might not be widely accepted by management ... or developers with a high need for job security.

Reviewer 4: It might also make sense to mention to not be too strict about what can be done with the SW. We (at Bosch) made that mistake and precluded the use of InnerSource software in Open Source projects later on. This turned out to be a killer criterium for some of our projects and eventually triggered the development of a new version of the license, which took a lot of effort.

DB Systel created their own InnerSource License, see [DB Inner Source License][db-inner-source-license]. They used the [EUPL][eupl], as that offered an open-source-like starting point, and then worked out the constraints and additional rules required in their specific organizational context.

The first legal entities (companies) within the DB AG are using their InnerSource License.

One positive effect that is already showing is that it simplifies the conversation, especially if some of the involved parties don’t know the InnerSource concept that well yet. Licenses are a well-known concept, therefore having an InnerSource License is a great discussion starter.

The experiments are also uncovering that there are further collaboration challenges that need to be solved in order to lead to a true InnerSource contribution and collaboration model.

The mentioned collaboration challenges include:

- making InnerSource licensed projects discoverable
- building communities for collaboration on projects, just like in Open Source

It is worth mentioning that so far the software shared under this InnerSource license is mostly tooling, infrastructure, and tools lower in the stack.

> 

Reviewer 4: Add Bosch as an known instance, too. Here's my proposal for content:

"The InnerSource initiative at Robert Bosch GmbH was also governed with a license from the beginning in 2009. With over 500 legal entities, sharing software internally was very difficult and time consuming to implement in a compliant way. InnerSource and specifically the InnerSource license changed that completely. Today, collaboration in software projects governed by their InnerSource license is the de facto standard and widely practiced. Bosch has evolved their license over the years, e. g. to also allow the eventual publication of InnerSource software as Open Source software, which wasn't possible with early versions of their license."
